### ISSUE TITLE
Race condition: Upgrades may run simultaneously.

### ISSUE THREAD

**Issue by sybrew** [sybrew/the-seo-framework#463]
Two upgrade processes may be invoked simultaneously. This means that, when two requests come in at the same time, the upgrader will run TWICE. 

This issue can happen on busy sites, where two requests come in at a very short time, and TSF will then try to upgrade the site twice.

This isn't a big deal for most of the upgrading process. But, some options are based on changing the option non-destructively, persistently. For example, the `home_title_location` is flipped from `left` to `right`, and vice versa. When this happens twice, the end result is that the option isn't flipped.

We need to find a way to lock this function without locking users out of their site, and with zero chance of bricking the site. The lock must be lifted whenever the site hangs, crashes, or becomes unavailable (server restart).

Using WordPress cronjobs may seem like a good solution, but the user may change the settings before such a job runs. This might result in the options being changed before the upgrader has a chance to run (causing unexpected outcomes), and the site may experience issues or erratic data output in the meantime.

We planned to release v4.0 today, but this bug is a blocker and prevents us from doing so.

**Comment by sybrew** [comment]
These conditions apply in preventing unexpected behavior:

1. The upgrader process must be able to run on the front-end. This is because we can't expect the site owner to be present when TSF updates (like on multisites...).
2. The upgrader process must always be able to complete before the user can interact with the options (this is already the case, but we may not want to destroy that).
3. The locker must clear on failure. We know that WordPress 5.2 has a "site crashed" mailer-system, we may want to borrow that behavior. We already allow for the upgrader to continue where it left off, but this doesn't pertain to individual database interactions, only version-upgrades.
4. The locker and upgrader functions MUST flush the site's cache. Some caching plugins may interfere, which we've previously resolved in v3.1.4 update.
5. The locker must lock out the user from accessing the SEO settings. We also have a redirect in place that prevents users from accessing the options when the upgrader is running&mdash;this must be persistent.

**Comment by sybrew** [comment]
We're going to opt for an auto-releasing lock that expires after 300 seconds. After which, the upgrade will retry.

This lock is going to rely on a self-expiring WordPress database transient:
```php
if ( get_transient( 'tsf_upgrade_lock' ) ) return;
set_transient( 'tsf_upgrade_lock', true, 300 );
register_shutdown_function( __NAMESPACE__ . '\\_release_upgrade_lock' );
```

This is not the best mitigation against all race conditions. For example, it won't prevent the microsecond-issues unless the database has a locking mechanism in place: "we're writing now... please hold until we release it...". Even then, how will the database know it's locked? Only the precision-level of high-performance CPUs abiding by a von Neumann process will be able to tell. PHP can't come close to that level of precision... ever.

That said, however, this lock will prevent the millisecond (and beyond) issues, which are most prone to happen.

The best mitigation would be to prepare a future (in 10 seconds or later), single-running, cronjob, with a transient attached as it's running. Unfortunately, this will cause a bad user experience, as we then must mitigate user-error for 99.9% of cases that don't require this level of precision.

With that, I consider this fixed in the upcoming commit.
